Spigot-Server-Patches/0731-additions-to-PlayerGameModeChangeEvent.patch

```diff
@@ -0,0 +1,135 @@
++                    if (event != null && event.isCancelled()) {
++                        commandlistenerwrapper.sendMessage(io.papermc.paper.adventure.PaperAdventure.asVanilla(event.cancelMessage()), false);
++                        continue;
++                    }
++                    // Paper end
+                     ++i;
+                 }
+             }
+diff --git a/src/main/java/net/minecraft/server/level/EntityPlayer.java b/src/main/java/net/minecraft/server/level/EntityPlayer.java
+index 10e9e5328f783832b957113a8672f45f90ace813..c1184eafaf9931e501943282ae609f753181c75d 100644
+--- a/src/main/java/net/minecraft/server/level/EntityPlayer.java
++++ b/src/main/java/net/minecraft/server/level/EntityPlayer.java
+@@ -395,7 +395,14 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
+         if (this.locY() > 300) this.setPositionRaw(locX(), 257, locZ()); // Paper - bring down to a saner Y level if out of world
+         if (nbttagcompound.hasKeyOfType("playerGameType", 99)) {
+             if (this.getMinecraftServer().getForceGamemode()) {
++                // Paper start - call PlayerGameModeChangeEvent on join for players that do not have the correct gamemode
++                if (this.getMinecraftServer().getGamemode() != EnumGamemode.getById(nbttagcompound.getInt("playerGameType"))) {
++                    if (new org.bukkit.event.player.PlayerGameModeChangeEvent(this.getBukkitEntity(), GameMode.getByValue(this.getMinecraftServer().getGamemode().getId()), org.bukkit.event.player.PlayerGameModeChangeEvent.Cause.DEFAULT_GAMEMODE, null).callEvent()) {
+                 this.playerInteractManager.a(this.getMinecraftServer().getGamemode(), EnumGamemode.NOT_SET);
++                    } else {
++                        this.playerInteractManager.a(EnumGamemode.getById(nbttagcompound.getInt("playerGameType")), nbttagcompound.hasKeyOfType("previousPlayerGameType", 3) ? EnumGamemode.getById(nbttagcompound.getInt("previousPlayerGameType")) : EnumGamemode.NOT_SET); // copied from below; if cancelled, set gamemode normally
++                    }
++                } // Paper end
+             } else {
+                 this.playerInteractManager.a(EnumGamemode.getById(nbttagcompound.getInt("playerGameType")), nbttagcompound.hasKeyOfType("previousPlayerGameType", 3) ? EnumGamemode.getById(nbttagcompound.getInt("previousPlayerGameType")) : EnumGamemode.NOT_SET);
+             }
+@@ -1782,16 +1789,16 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
+     }
+ 
+     @Override
+-    public void a(EnumGamemode enumgamemode) {
++    public PlayerGameModeChangeEvent a(EnumGamemode enumgamemode, org.bukkit.event.player.PlayerGameModeChangeEvent.Cause cause, net.kyori.adventure.text.Component message) { // Paper - Add cause and nullable message to event
```

Comment:
WHEN WILL YOU LEARN?! WHEN WILL YOU LEARN?! WHEN WILL YOU LEARN THAT YOUR ACTIONS HAVE CONSEQUENCES?!

Reply:
Apparently never